tools/wake/config.cpp

```diff
@@ -0,0 +1,313 @@
+  constexpr size_t read_side = 0;
+  constexpr size_t write_side = 1;
+
+  // Setup a pipe for the child process to write to
+  int stdoutPipe[2];
+  if (pipe(stdoutPipe) < 0) {
+    perror("Failed to allocate eviction pipe");
+    exit(EXIT_FAILURE);
+  }
+
+  // Fork to get a child process
+  pid_t pid = fork();
+  if (pid < 0) {
+    perror("Failed to fork eviction process");
+    exit(EXIT_FAILURE);
+  }
+
+  if (pid == 0) {
+    // A slight quark is that spaces between seperate parts of `to_expand`
+    // will only have one space between them at the end of this. This may
+    // catch some people off guard but adding `double` quotes around this
+    // would require escaping things and would change expansion. Since
+    // the intended use cases will likely not have spaces in them this
+    // seems like an ok trade off.
+    std::string shell_string = "echo " + to_expand;
+
+    // overwrite stdout with the write side of the pipe
+    if (dup2(stdoutPipe[write_side], STDOUT_FILENO) == -1) {
+      perror("Failed to dup2 stdout pipe for eviction process");
+      exit(EXIT_FAILURE);
+    }
+
+    // We don't want stdin/stderr messing things up and
+    // we no longer need the write/read side of the pipe
+    // since stdout *is* the write side of the pipe now.
+    close(STDIN_FILENO);
+    // close(STDERR_FILENO);
+    close(stdoutPipe[read_side]);
+    close(stdoutPipe[write_side]);
+
+    // Now replace this process with the shell command but
+    // exit and fail if we don't manage to do that.
+    execl("/bin/sh", "/bin/sh", "-c", shell_string.c_str(), nullptr);
+    perror("Failed to exec /bin/sh");
+    exit(EXIT_FAILURE);
+  }
+
+  // Make sure to close this side of the pipe so that the shell program doesn't
+  // stay open waiting for commands forever.
+  close(stdoutPipe[write_side]);
+
+  // Read everything the child process outputs and append
+  // it to a string as we go.
+  std::string acc;
+  char buf[4096];
+  while (true) {
+    ssize_t size = read(stdoutPipe[read_side], buf, sizeof(buf));
+    if (size < 0 && errno == EINTR) {
+      continue;
+    }
+    if (size < 0) {
+      perror("failed to exec /bin/sh");
+      exit(EXIT_FAILURE);
+    }
+    if (size == 0) {
+      break;
+    }
+    acc.append(buf, buf + size);
+  }
+
+  // We should be done already but let's wait for the child and
+  // make sure everything worked well there.
+  int status = 1;
+  if (waitpid(pid, &status, 0) != pid) {
+    perror("waitpid: failed to wait /bin/sh -c echo ...");
+    exit(EXIT_FAILURE);
+  }
+  if (status != 0) {
+    fprintf(stderr, "waitpid: /bin/sh -c echo ... failed with non-zero exit status");
+    exit(EXIT_FAILURE);
+  }
+
+  // Remove the newline added by echo
+  // echo -n doesn't work MacOS
+  acc.pop_back();
+
+  // Return the string we got back
+  return acc;
+}
+
+// Find the default location for the user level wake config
+static std::string default_user_config() {
```

Comment:
The semantics should be a bit different in this function. If XDG_CONFIG_HOME is set then we don't need HOME to be set. Also HOME only needs to be set if the user did not provide us with a user config path. You can either return an optional here when there is no valid default, or you can delay setting the default until after you've checked for the user config. Or you can make this function take an optional from expect_string and have it contain all the logic for setting one thing or another.

Reply:
I personally don't think all that is necessary. I think its fine for us to demand that HOME always be set and the above is a large amount of extra code/maintenance burden just to not demand HOME be set in a couple of cases

Reply:
Talked in person with Jake, skipping this comment for now. We can review if we find ourselves in a position where we can't demand HOME be set